### PR TITLE
Cleanup compass probing

### PR DESCRIPTION
**Pull request by tridge**
This switches Linux compass probing over to the new probe list method, which allows us to remove the horrible ifdef tree in AP_Compass.cpp
This also fixes a bug where we could probe the same i2c compass twice (that bug was introduced recently)

### CODE REVIEW

**Comment by tridge** [comment]
@OXINARF can you have a look at this one?

**Comment by tridge** [comment]
After discussion with Peter and Randy I will merge this. The change to fix the duplicate compasses showing up on i2c is pretty important to get in now

**Comment by tridge** [comment]
@OXINARF I'd still appreciate some post-merge review. Note that the changes in linux.h quite deliberately don't line up with the old compass probing in AP_Compass. It turns out that a lot of the Linux boards had broken compass config already, for example specifying external i2c compasses in the AP_Compass probe but also doing external probing

File: libraries/AP_HAL/board/linux.h [hunk 1]
```diff
@@ -244,18 +229,15 @@
...
     #define HAL_INS_PROBE_LIST PROBE_IMU_SPI(BMI160, "bmi160")
     #define HAL_BARO_PROBE_LIST PROBE_BARO_I2C(MS56XX, 2, 0x76)
-    #define HAL_COMPASS_DEFAULT HAL_COMPASS_AERO
-    #define HAL_COMPASS_BMM150_I2C_BUS 2
-    #define HAL_COMPASS_BMM150_I2C_ADDR 0x12
-    #define HAL_COMPASS_IST8310_I2C_BUS 4
-    #define HAL_COMPASS_IST8310_I2C_ADDR 0x0E
-    #define HAL_COMPASS_HMC5843_I2C_BUS 4
-    #define HAL_COMPASS_HMC5843_I2C_ADDR 0x1E
+    #define HAL_MAG_PROBE1 PROBE_MAG_I2C(BMM150, 2, 0x12, ROTATION_NONE)
```
Comment (OXINARF) [inline libraries/AP_HAL/board/linux.h]: BMM150 doesn't have a rotation argument.

File: libraries/AP_HAL/board/linux.h [hunk 2]
```diff
@@ -289,8 +268,7 @@
 #elif CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_RST_ZYNQ
     #define HAL_INS_PROBE_LIST PROBE_IMU_SPI2(Invensense, "rst_g", "rst_a", ROTATION_ROLL_180_YAW_90, ROTATION_ROLL_180_YAW_90)
     #define HAL_BARO_PROBE_LIST PROBE_BARO_SPI(MS56XX, "ms5611")
-    #define HAL_COMPASS_DEFAULT HAL_COMPASS_LIS3MDL
-    #define HAL_COMPASS_LIS3MDL_NAME "lis3mdl"
+    #define HAL_MAG_PROBE_LIST PROBE_MAG_SPI(LIS3MDL, lis3mdl, false, ROTATION_ROLL_180_YAW_90)
```
Comment (OXINARF) [inline libraries/AP_HAL/board/linux.h]: lis3mdl needs to have quotes, it needs to be a string.

File: libraries/AP_Baro/AP_Baro.cpp [hunk 3]
```diff
@@ -473,6 +473,7 @@ void AP_Baro::init(void)
     }
 #endif
 
+// macro for use by HAL_INS_PROBE_LIST
```
Comment (OXINARF) [inline libraries/AP_Baro/AP_Baro.cpp]: HAL_BARO_PROBE_LIST

File: libraries/AP_HAL_ChibiOS/hwdef/skyviper-f412-rev1/hwdef.dat [hunk 4]
```diff
@@ -129,9 +129,7 @@ define LAND_DETECTOR_ACCEL_MAX 2.0f
 
 define CHIBIOS_ADC_MAVLINK_DEBUG 1
 
-define HAL_COMPASS_DEFAULT HAL_COMPASS_BMM150_I2C
-define HAL_COMPASS_BMM150_I2C_BUS   0
-define HAL_COMPASS_BMM150_I2C_ADDR  0x10
+COMPASS BMM150 I2C:0:0x10 ROTATION_NONE
```
Comment (OXINARF) [inline libraries/AP_HAL_ChibiOS/hwdef/skyviper-f412-rev1/hwdef.dat]: As said, BMM150 doesn't have rotation argument.

File: libraries/AP_HAL_ChibiOS/hwdef/skyviper-f412/hwdef.dat [hunk 5]
```diff
@@ -120,9 +120,7 @@ define ALLOW_ARM_NO_COMPASS
 
 define CHIBIOS_ADC_MAVLINK_DEBUG 1
 
-define HAL_COMPASS_DEFAULT HAL_COMPASS_BMM150_I2C
-define HAL_COMPASS_BMM150_I2C_BUS   0
-define HAL_COMPASS_BMM150_I2C_ADDR  0x10
+COMPASS BMM150 I2C:0:0x10 ROTATION_NONE
```
Comment (OXINARF) [inline libraries/AP_HAL_ChibiOS/hwdef/skyviper-f412/hwdef.dat]: Same.

File: libraries/AP_HAL/board/linux.h [hunk 6]
```diff
@@ -8,13 +8,18 @@
 #define HAL_STORAGE_SIZE_AVAILABLE  HAL_STORAGE_SIZE
 
 // make sensor selection clearer
-#define PROBE_IMU_I2C(driver, bus, addr, args ...) ADD_BACKEND(AP_InertialSensor_ ## driver::probe(*this,hal.i2c_mgr->get_device(bus, addr),##args))
+#define PROBE_IMU_I2C(driver, bus, addr, args ...) ADD_BACKEND(AP_InertialSensor_ ## driver::probe(*this,GET_I2C_DEVICE(bus, addr),##args))
```
Comment (OXINARF) [inline libraries/AP_HAL/board/linux.h]: Pointing out this one, but it's also for others below: this is missing the `std::move` call.

Side comment: I **think** this is probably fine and I would even go further: delete the copy assignment operator from OwnPtr and we should be able to remove most std::move calls - the compiler should then do a copy/move elision (which the std::move call prevents).

**Comment by patrickpoirier51** [comment]
@oximarf on BeagleBond I2C bus 0 is for power management unit an bus 1 is used to read capes eeprom configuration , probing these bus is not recommended

**Comment by OXINARF** [comment]
@patrickpoirier51 Can you test if BB boards are running well with master?

**Comment by patrickpoirier51** [comment]
OK I will test later today and report

**Comment by patrickpoirier51** [comment]
OK Tested (static test) both on BBBMINI and BBBlue , no issue

**Pull request merged**